model/riscv_vmem_pte.sail

```diff
@@ -51,10 +58,33 @@ function pte_is_ptr(pte_flags : PTE_Flags) -> bool = (pte_flags[X] == 0b0)
                                                       & (pte_flags[W] == 0b0)
                                                       & (pte_flags[R] == 0b0)
 
+// Extension hooks can request standard PTE validity checks by returning Ext_PTE_Std
+union Ext_isInvalPTE_Check = {
+  Ext_PTE_Invalid : unit,
+  Ext_PTE_Valid   : unit,
+  Ext_PTE_Std     : unit
+}
+
+// Extension hook to override standard definition of extended PTE attribute bits
+function ext_pte_is_invalid (p : PTE_Flags, e : PTE_Ext_Flags) -> Ext_isInvalPTE_Check =
+  Ext_PTE_Std()
+
 // PRIVATE: check if a PTE is valid
-function pte_is_invalid(pte_flags : PTE_Flags) -> bool = (pte_flags[V] == 0b0)
-                                                         | ((pte_flags[W] == 0b1)
-                                                            & (pte_flags[R] == 0b0))
+function pte_is_invalid(pte_flags : PTE_Flags, pte_ext_flags : PTE_Ext_Flags) -> bool = {
+  match ext_pte_is_invalid(pte_flags, pte_ext_flags) {
+    Ext_PTE_Invalid() => true,
+    Ext_PTE_Valid()   => false,
+    Ext_PTE_Std()     => {
+      (pte_flags[V] == 0b0) |
+      (pte_flags[W] == 0b1 & pte_flags[R] == 0b0) |
+      (pte_ext_flags[RSVD] != 0b0000000 | pte_ext_flags[PBMT] == 0b11) |
+      (not(pte_is_ptr(pte_flags)) & (pte_ext_flags[N] != 0b0 ) & (not(haveSvnapot()))) |
+      (not(pte_is_ptr(pte_flags)) & (pte_ext_flags[PBMT] != 0b00) & (not(haveSvpbmt()) | menvcfg[PBMTE] == 0b0)) |
+      (pte_is_ptr(pte_flags) & (pte_flags[D] == 0b1 | pte_flags[A] == 0b1 | pte_flags[U] == 0b1 |
+                                 pte_ext_flags[N] != 0b0 | pte_ext_flags[PBMT] != 0b00))
+    }
+  }
+}
```

Comment:
I think this can be simplified a bit. Here's our version, IIRC I based it on yours.

```
function pte_is_invalid(pte_flags : PTE_Flags, pte_ext : PTE_Ext) -> bool =
  match ext_pte_validity(pte_flags, pte_ext) {
    Ext_PTE_Invalid => true,
    Ext_PTE_Valid   => false,
    Ext_PTE_Std     =>
        pte_flags[V] == 0b0
      | (pte_flags[W] == 0b1 & pte_flags[R] == 0b0)
      // These bits must be zero unless the relevant extension is supported
      // and this is a leaf node.
      | pte_ext[N]    != zeros() & (not(haveSvnapot()) | pte_is_ptr(pte_flags))
      | pte_ext[PBMT] != zeros() & (not(haveSvpbmt())  | pte_is_ptr(pte_flags))
      // PBMT 0b11 is always reserved. This applies even without PBMT support.
      | pte_ext[PBMT] == 0b11
      // Reserved bits must be 0.
      | pte_ext[reserved] != zeros()
  }
```

Reply:
This version is missing a few things:
1. A, D, and U flags are reserved in non-leaf PTEs
2. PBMT is reserved if menvcfg.PBMTE is 0
3. N bit is reserved for superpage leaf PTEs

Reply:
updated.

Reply:
@Timmmm - let me know if any other changes are needed.